File: 4949.py
import sys
import re

com = re.compile('[a-zA-Z]')

story = []
while 1:
    x = sys.stdin.readline().rstrip()
    story.append(x)
    if x == '.':
        break

def vps(line):
    stack = []
    if line[0] == ")" or line[0] == "]":
        # print("1")
        return "no"
    else: pass
    for s in line:
        if bool(com.match(s)) == True:
            pass
        elif s == " " or s == ".":
            pass
        elif s == "(" or s == "[":
            stack.append(s)
            # print(stack)
        elif s == ")" and stack != []:
            if stack[-1] == "(":
                stack.pop()
            else:
                # print("2")
                return "no"
        elif s == "]" and stack != []:
            if stack[-1] == "[":
                stack.pop()
            else:
                # print("3")
                return "no"
        else:
            # print("4")
            return "no"
    if stack == []:
        return "yes"
    else:
        # print("5")
        return "no"

for i in range(len(story)-1):
    result = vps(story[i])
    print(result)

File: Go_solutions/4949.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func vps(x string) string {
	var stack []string

	// fmt.Println(string(x[0]))
	// return x
	if string(x[0]) == ")" || string(x[0]) == "]" {
		return "no"
	}
	for _, ch := range x {
		s := string(ch)
		match, _ := regexp.MatchString("[a-zA-Z]", s)
		if match == true {
		} else if s == " " || s == "." {
		} else if s == "(" || s == "[" {
			stack = append(stack, s)
		} else if s == ")" && stack != nil {
			if stack[len(stack)-1] == "(" {
				if len(stack) == 1 {
					stack = nil
				} else {
					stack = stack[:len(stack)-1]
				}
			} else {
				return "no"
			}
		} else if s == "]" && stack != nil {
			if stack[len(stack)-1] == "[" {
				if len(stack) == 1 {
					stack = nil
				} else {
					stack = stack[:len(stack)-1]
				}
			} else {
				return "no"
			}
		} else {
			return "no"
		}
	}
	if stack == nil {
		return "yes"
	} else {
		return "no"
	}
}

func main() {
	inputReader := bufio.NewReader(os.Stdin)
	var story []string
	for {
		input, _ := inputReader.ReadString('\n')
		action := strings.TrimSpace(input)
		// fmt.Printf("%d\n", len(input))
		story = append(story, action)

		// fmt.Println(story)

		if string(input[0]) != " " && strings.TrimSpace(input) == "." {
			break
		}
		// fmt.Println(story)
	}
	// fmt.Println(story)
	for _, line := range story[:len(story)-1] {
		answer := vps(line)
		fmt.Println(answer)
	}
}
